### ISSUE TITLE
Structural parameter leads to unknown array size, even though the size does not depend on the parameter 

### ISSUE THREAD

**Issue by FWuellhorst** [OpenModelica/OpenModelica#12106]
### Description

Having a structural parameter in the model `IBPSA.Fluid.Movers.FlowControlled_m_flow` leads to errors in OpenModelica, while other tools (Dymola, Optimica) work just fine.
Error:
```
[IBPSA.Fluid.Movers.BaseClasses.Characteristics.efficiencyParameters_yMot: 5:3-6:45]: [Could not evaluate structural parameter (or constant): moverExampleStructual.pump_m_flow.eff.per.pressure.V_flow which gives dimensions of array: y. Array dimensions must be known at compile time.](omeditmessagesbrowser:///IBPSA.Fluid.Movers.BaseClasses.Characteristics.efficiencyParameters_yMot?lineNumber=5)
```

The array size is fixed in `IBPSA.Fluid.Movers.Data.Generic` for both usages of `IBPSA.Fluid.Movers.BaseClasses.Characteristics.efficiencyParameters_yMot`:
```
  parameter
    IBPSA.Fluid.Movers.BaseClasses.Characteristics.efficiencyParameters_yMot
    motorEfficiency_yMot(y={0}, eta={0.7});
  final parameter
    IBPSA.Fluid.Movers.BaseClasses.Characteristics.efficiencyParameters_yMot
      motorEfficiency_yMot_generic=
        IBPSA.Fluid.Movers.BaseClasses.Characteristics.motorEfficiencyCurve(
          P_nominal=WMot_nominal,
          eta_max=etaMot_max);
```
with `IBPSA.Fluid.Movers.BaseClasses.Characteristics.motorEfficiencyCurve`:
```
function motorEfficiencyCurve
  "This function generates the generic curve for motor efficiency based on rated motor power"
  extends Modelica.Icons.Function;
  input Modelica.Units.SI.Power P_nominal "Rated input power of the motor";
  input Modelica.Units.SI.Efficiency eta_max "Maximum motor efficiency";
  // Below, default values are assigned so that that dimensions are known.
  output IBPSA.Fluid.Movers.BaseClasses.Characteristics.efficiencyParameters_yMot
    motorEfficiency_yMot(
        y={0, 0.1, 0.2, 0.3, 0.4, 0.6, 0.8, 1, 1.2},
        eta=zeros(nPoi)) "Motor efficiency vs. motor part load ratio";
```
Thus, sizes should always be 1 and 9. 

### Steps to Reproduce

```
model ArraySizeError
  IBPSA.Fluid.Movers.Examples.MoverParameter moverExampleStructual(m_flow_nominal=m_flow_nominal);
  parameter Real m_flow_nominal annotation(Evaluate = false);
end ArraySizeError;
```
I am aware that giving m_flow_nominal solves the issue, but setting the value to the output of another function (like `Modelica.Blocks.Tables.Internal.getTable2DValueNoDer2`) raises the same error. 
(The model `IBPSA.Fluid.HeatPumps.ModularReversible.Examples.ReversibleAirToWater2D_OneRoomRadiator` in the [branch ](https://github.com/ibpsa/modelica-ibpsa/tree/issue1576_heatPumpModelIntegration) provides an example of the error.

### Expected Behavior

No error.

### Version and OS

 - OpenModelica Version: OpenModelica v1.23.0-dev-375-gf02e3af233 (64-bit)
Conne
 - OS: Windows 10, 64 bit
 - Versions of used Modelica libraries if applicable: https://github.com/ibpsa/modelica-ibpsa/tree/issue1576_heatPumpModelIntegration (same problem on master)

### Additional Context

https://github.com/ibpsa/modelica-ibpsa/pull/1628

**Comment by perost** [comment]
#12131 should fix the reported error, but unfortunately it doesn't seem to be enough since it now fails with:
```
[IBPSA/Fluid/HeatPumps/ModularReversible/ReversibleAirToWaterTableData2D.mo:74:3-74:171:writable] Warning: In relation heaPum.refCyc.refCycHeaPumHea.PEle_nominal <> heaPum.refCyc.refCycHeaPumCoo.PEle_nominal, <> on Real numbers is only allowed inside functions.
[NFCeval.mo:1134:9-1134:67:writable] Error: Internal error NFCeval.evalBinaryDiv failed to evaluate ‘2e4 / heaPum.refCyc.refCycHeaPumHea.QHeaNoSca_flow_nominal‘
[NFCeval.mo:2044:20-2044:79:writable] Error: Internal error NFCeval.evalBuiltinAbs got invalid arguments (pumHeaPum.eff.per.pressure.dp[pumHeaPum.eff.nOri])
```
The warning about `<>` isn't the problem in this case, but maybe something you want to fix anyway. But there seems to still be some issues with evaluating parameters, possibly unrelated to the original issue from the looks of it.

**Comment by casella** [comment]
@perost in any case maybe it's better if you close this ticket and open another one with the new issue. Thanks!

**Comment by FWuellhorst** [comment]
Thanks for the quick fix!
@perost: The <> was just a temporary test, I removed it again. 
@casella @perost : Maybe I can help with the debugging of `2e4 / heaPum.refCyc.refCycHeaPumHea.QHeaNoSca_flow_nominal`:
`QHeaNoSca_flow_nominal` is defined in a replaceable model like this: 
```
  parameter Modelica.Units.SI.HeatFlowRate QHeaNoSca_flow_nominal=Modelica.Blocks.Tables.Internal.getTable2DValueNoDer2(
        tabIdeQUse_flow,
        TCon_nominal,
        TEva_nominal)
    "Unscaled nominal heating capacity "
    annotation (Dialog(group="Nominal condition"));
 
   parameter Modelica.Blocks.Types.ExternalCombiTable2D tabIdeQUse_flow=
      Modelica.Blocks.Types.ExternalCombiTable2D(
      "NoName",
      "NoName",
      tabQUse_flow.table,
      smoothness,
      extrapolation,
      false) "External table object for nominal useful side conditions";
 ```
 Where the table is a replaceable parameter record. Both `TCon_nominal` and `TEva_nominal` are not structural. 
 Dymola throws warnings about non-literal values, as it seems to non evaluate the table call during model check:
 ```
  Non-literal value.

In nominal attribute for pumHeaPum.m_flow_actual.
Ignoring nominal attribute.
In class Modelica.Blocks.Interfaces.
Near file: C:/Program Files/Dymola 2023x/Modelica/Library/Modelica 4.0.0/Blocks/Interfaces.mo, line 2
Used in component pumHeaPum.
Near file: D:/04_git/modelica-ibpsa/IBPSA/Fluid/HeatPumps/ModularReversible/Examples/BaseClasses/PartialOneRoomRadiator.mo, line 74
```
However, it simulates just fine.

**Comment by perost** [comment]
> What's wrong with `2e4 / heaPum.refCyc.refCycHeaPumHea.QHeaNoSca_flow_nominal`?

`NFCeval.evalBinaryDiv` expects both operands to already have been constant evaluated, which is evidently not the case here.

> @perost please check this [regression report](https://libraries.openmodelica.org/branches/history/master/2024-03-22%2011:29:51..2024-03-22%2023:42:01.html). Apparently #12131 commit broke `ModelicaTest.Math.TestMatrices2` and ModelicaTest.Math.TestVectors, which now fail in the frontend because of issues with dimension inference.
> 
> There are also a few regressions on Buildings models, which now fail during Templates, not sure if #12131 is to blame, maybe you can check.
> 
> Thanks!

The ModelicaTest regressions should be fixed by #12136. For the Buildings ones I'm actually not sure, I can't replicate that issue with either Buildings from our package manager or from their git but it doesn't look like something that would be caused by my changes. I'll take responsibility for the one Buildings model that went from failed to verified though.

**Comment by casella** [comment]
> The ModelicaTest regressions should be fixed by #12136.

Great, thanks! 😃 

> For the Buildings ones I'm actually not sure, I can't replicate that issue with either Buildings from our package manager or from their git but it doesn't look like something that would be caused by my changes.

I agree, I think it's changes that were applied to the Buildings library in PR lbl-srg/modelica-buildings#3709. I opened lbl-srg/modelica-buildings#3725, let's see.

> I think it' I'll take responsibility for the one Buildings model that went from failed to verified though.

Absolutely 😄

**Comment by perost** [comment]
The issue with failing to evaluate some parameters should be fixed with #12181, and the model now simulates (though with some warnings).

The issue was that we try hard to split binding equations for record component so we don't have to move them to an initial equation, but when there was an if-expression in the binding equation we would try to evaluate the condition of it and then marked the condition as structural regardless of whether the evaluation succeeded or not. Now we check that the condition was evaluated before marking it as structural, so we don't mark any parameters as structural that can't actually be evaluated later.

**Comment by FWuellhorst** [comment]
@perost Nice, the nightly build works on my side, as well. Thank you very much for fixing this! The remaining warnings are wanted.
